File: ConcurrentDispatchWithNoDependency.m
#import "ConcurrentDispatchWithNoDependency.h"
#import "AsyncDispatch.h"

void
StartConcurrentDispatchWithNoDependency(ReplayContext *context)
{
	assert(context->concurrent);
	StartAsyncDispatch(context->councurrencyLimit);
}

void
FinishConcurrentDispatchWithNoDependencyAndWait(ReplayContext *context)
{
	FinishAsyncDispatchAndWait();
}

void
DispatchTasksConcurrentlyWithNoDependency(NSArray<NSDictionary*> *playlist, ReplayContext *context)
{
	StartConcurrentDispatchWithNoDependency(context);

#if TRACE
	printf("start dispatching async tasks\n");
#endif

	Class dictionaryClass = [NSDictionary class];

	for(id oneStep in playlist)
	{
		if([oneStep isKindOfClass:dictionaryClass])
		{
			HandleActionStep((NSDictionary *)oneStep, context,
				^(dispatch_block_t action,
				__unused NSArray<NSString*> *inputs,
				__unused NSArray<NSString*> *exclusiveInputs,
				__unused NSArray<NSString*> *outputs)
				{
					if(action != NULL)
					{
						AsyncDispatch(action);
					}
				});
		}
		else
		{
			fprintf(gLogErr, "error: invalid non-dictionary step in the playlist\n");
		}
	}

#if TRACE
	printf("done dispatching async tasks\n");
#endif

	FinishConcurrentDispatchWithNoDependencyAndWait(context);
}

void
DispatchTaskConcurrentlyWithNoDependency(NSDictionary *stepDescription, ReplayContext *context)
{
	HandleActionStep(stepDescription, context,
		^(dispatch_block_t action,
		__unused NSArray<NSString*> *inputs,
		__unused NSArray<NSString*> *exclusiveInputs,
		__unused NSArray<NSString*> *outputs)
		{
			if(action != NULL)
			{
				AsyncDispatch(action);
			}
		});
}

File: replay/ConcurrentDispatchWithNoDependency.cpp
#include "ConcurrentDispatchWithNoDependency.h"
#include "AsyncDispatch.h"

void
StartConcurrentDispatchWithNoDependency(ReplayContext *context)
{
	assert(context->concurrent);
	StartAsyncDispatch(context->councurrencyLimit);
}

void
FinishConcurrentDispatchWithNoDependencyAndWait(ReplayContext *context)
{
	FinishAsyncDispatchAndWait();
}

void
DispatchTasksConcurrentlyWithNoDependency(const std::vector<ActionStep>& playlist, ReplayContext *context)
{
	StartConcurrentDispatchWithNoDependency(context);

#if TRACE
	printf("start dispatching async tasks\n");
#endif

	for (const auto& step : playlist)
	{
		HandleActionStep(step, context,
			[](std::function<void()> action,
			__unused std::vector<std::string> inputs,
			__unused std::vector<std::string> mutatingInputs,
			__unused std::vector<std::string> exclusiveInputs,
			__unused std::vector<std::string> outputs)
			{
				if(action)
					AsyncDispatch(std::move(action));
			});
	}

#if TRACE
	printf("done dispatching async tasks\n");
#endif

	FinishConcurrentDispatchWithNoDependencyAndWait(context);
}

void
DispatchTaskConcurrentlyWithNoDependency(ActionStep step, ReplayContext *context)
{
	HandleActionStep(std::move(step), context,
		[](std::function<void()> action,
		__unused std::vector<std::string> inputs,
		__unused std::vector<std::string> mutatingInputs,
		__unused std::vector<std::string> exclusiveInputs,
		__unused std::vector<std::string> outputs)
		{
			if(action)
				AsyncDispatch(std::move(action));
		});
}
